File: src/modifiers/handGestureShortcut.ts
import { createModifier } from './createModifier';

/// Which system hand-gesture (double-tap / pinch) action a view opts into.
/// `primaryAction` maps to `HandGestureShortcut.primaryAction` (treat as the
/// view's primary action); `counter` maps to `HandGestureShortcut.counter`.
export type HandGestureShortcutKind = 'primaryAction' | 'counter';

export interface HandGestureShortcutParams {
  /// The hand-gesture action this view should respond to.
  shortcut: HandGestureShortcutKind;
}

/// SwiftUI `.handGestureShortcut(_:)` — assigns a view (typically a `Button`)
/// to a system hand-gesture shortcut so it can be triggered by the Apple Watch
/// double-tap / pinch gesture. Requires watchOS 10; on older systems the view
/// is returned unchanged.
export function handGestureShortcut(
  shortcut: HandGestureShortcutKind
): ReturnType<typeof createModifier>;
export function handGestureShortcut(
  params: HandGestureShortcutParams
): ReturnType<typeof createModifier>;
export function handGestureShortcut(
  arg: HandGestureShortcutKind | HandGestureShortcutParams
) {
  if (typeof arg === 'string') {
    return createModifier('handGestureShortcut', { shortcut: arg });
  }
  return createModifier('handGestureShortcut', arg);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/HandGestureShortcut.swift
import SwiftUI

/// SwiftUI `.handGestureShortcut(_:)`. Assigns a view to a system
/// hand-gesture shortcut (the Apple Watch double-tap / pinch gesture).
/// Requires watchOS 11 — on older systems the view is returned unchanged.
///
/// `shortcut`: `.primaryAction` is the only hand-gesture shortcut on
/// watchOS, so every value maps to it (`.counter` exists only on other
/// platforms). The `shortcut` param is accepted for parity.
enum RNWHandGestureShortcutModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("handGestureShortcut") { view, _, _ in
            guard #available(watchOS 11.0, *) else { return view }
            return AnyView(view.handGestureShortcut(.primaryAction))
        }
    }
}
